snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -0,0 +1,182 @@
+    return FunctionCall(None, function_name, arguments)
+
+
+METRICS_SET_TYPE = "s"
+METRICS_DISTRIBUTIONS_TYPE = "d"
+METRICS_COUNTERS_TYPE = "c"
+
+
+class MetricsAggregateProcessor(MessageProcessor, ABC):
+    TEN_SECONDS = 10
+    ONE_MINUTE = 60
+    ONE_HOUR = 3600
+    ONE_DAY = 3600 * 24
+    GRANULARITIES_SECONDS = [TEN_SECONDS, ONE_MINUTE, ONE_HOUR, ONE_DAY]
+
+    @abstractmethod
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    def process_message(
+        self, message: Mapping[str, Any], metadata: KafkaMessageMetadata
+    ) -> Optional[ProcessedMessage]:
+        if not self._should_process(message):
+            return None
+
+        timestamp = _ensure_valid_date(datetime.utcfromtimestamp(message["timestamp"]))
+        assert timestamp is not None
+
+        keys = []
+        values = []
+        tags = message["tags"]
+        assert isinstance(tags, Mapping)
+        for key, value in sorted(tags.items()):
+            assert key.isdigit()
+            keys.append(int(key))
+            assert isinstance(value, int)
+            values.append(value)
+
+        processed = [
+            {
+                "org_id": _literal(message["org_id"]),
+                "project_id": _literal(message["project_id"]),
+                "metric_id": _literal(message["metric_id"]),
+                "timestamp": _call(
+                    "toStartOfInterval",
+                    (
+                        _call("toDateTime", (_literal(timestamp.isoformat()),),),
+                        _call("toIntervalSecond", (_literal(granularity),)),
+                    ),
+                ),
+                "tags.key": _array_literal(keys),
+                "tags.value": _array_literal(values),
+                **self._process_values(message),
+                "retention_days": _literal(message["retention_days"]),
+                "partition": metadata.partition,
+                "offset": metadata.offset,
+                "granularity": _literal(granularity),
+            }
+            for granularity in self.GRANULARITIES_SECONDS
+        ]
+        return AggregateInsertBatch(processed, None)
+
+
+class SetsAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return (
+            settings.WRITE_METRICS_AGG_DIRECTLY
+            and message["type"] is not None
+            and message["type"] == METRICS_SET_TYPE
+        )
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        values = message["value"]
+        for v in values:
+            assert isinstance(v, int), "Illegal value in set. Int expected: {v}"
+        snuba_metrics.gauge("set.size", len(values))
```

Comment:
A gauge is not the metric you want to use.
A gauge is used to measure a single value that grows and shrinks. Example the concurrent queries on Clickhouse.
Here you want a counter. Each message logs the number of values so that we can later extract avg and quantiles, or just the aggregated sum of all messages to know how many values we process per minute/second etc.

`snuba_metrics.increment("set.size", len(values))`